Account Sync should happen at login, 1x per day, OR on-demand

mattia-malnis opened the pull request:
Closes #585 

To handle the account sync I've created a `can_sync?` method that:

- Skips sync if the account is not active
- Skips sync if the account is syncing in background;
- Permits account sync if `last_sync_date` is blank or is not today

There are a couple of things that I'd like to change but I don't know if this PR the right place:

### Sync account on activation/deactivation

Currently, we force an account sync in `AccountsController#update` when `account_params[:is_active] == "1"`.

https://github.com/maybe-finance/maybe/blob/ea3ba4f33a8fa06098489e7e6b34d56b38fff3e1/app/controllers/accounts_controller.rb#L21-L25

It's probably better to move this logic to an `after_save` callback, so we're always sure that account sync is triggered when the account is activated or deactivated.

### Show sync message when user clicks the "sync" button

Currently, the sync message is always shown because we pass the variable `is_syncing: true`:

https://github.com/maybe-finance/maybe/blob/ea3ba4f33a8fa06098489e7e6b34d56b38fff3e1/app/controllers/accounts_controller.rb#L56-L69

I'd simply remove the rendering of `accounts/sync_message` and let Action Cable render it when the job in the background starts.

zachgoll commented:
@mattia-malnis I'll start by addressing the two comments in the description (which you're welcome to address in this PR as I think they're related enough):

RE: Sync on activation/deactivation - the thought here was that if an account becomes deactivated, we no longer need to worry about syncing it.  But I'd agree that the existing logic could be improved.

RE: Show sync message when user clicks "sync" - I agree with your idea here.  That's a much simpler flow.

Will wait until the PR is marked ready to do a full code review.

mattia-malnis commented:
@zachgoll I've updated the code following your advice and the PR is now ready for review. 

While I was updating `AccountsController#sync`, I realized that there might be a case where the new logic we've implemented won't take effect as expected.

If the queue has many jobs and the user clicks the "sync" button 3 times in a row, we'll execute 3 separate jobs. This is because the account status is changed inside the `sync` method. A potential solution could be to move `update!(status: "syncing")` inside the `sync_later` method. However, I'm not sure if it's worth making this change now.

It's unlikely that a user would click the button multiple times in rapid succession since the notification hides the button for a couple of seconds after it's clicked. Nevertheless, I wanted to bring this potential issue to your attention.

zachgoll reviewed (commented):
@mattia-malnis I think the edge case you mentioned is okay to skip for now.  That will be an easy fix if it becomes problematic.

Overall, code looks good.  Just a few follow ups: 

- Family sync trigger (left a comment)
- Sync messages - your comment led me to think of a few additional requirements around this, so I'm going to open a separate issue here shortly to address a more comprehensive solution for this.

mattia-malnis commented on app/controllers/concerns/authentication.rb:
```diff
@@ -21,6 +21,10 @@ def user_signed_in?
     Current.user = user
+
+    # Sync accounts to ensure that exchange rates are up-to-date
+    Current.family.sync_accounts
```
Unfortunately, it won't work at the `application_controller` level because `Current.user` wil be always `nil` as is set up after in the child controllers with `before_action :authenticate_user!`:

https://github.com/maybe-finance/maybe/blob/b3c48d13e70780caa146cdb79fce6bb851fabd76/app/controllers/accounts_controller.rb#L1-L5

We should sync accounts after the user is found and before `last_login_at` is set with a new value, somenthing like: 

```ruby
module Authentication
  # ...

  def authenticate_user!
    if user = User.find_by(id: session[:user_id])
      Current.user = user

      if user.last_login_at.nil? || user.last_login_at < Date.current.beginning_of_day
        Current.family.sync_accounts
      end

      user.update(last_login_at: DateTime.now)
    else
      redirect_to new_session_url
    end
  end

  # ...
end
``` 

`authenticate_user!` is not the best place to put this logic, maybe it's more clear to move it at the controller level:

```ruby
class PagesController < ApplicationController
  before_action :authenticate_user!
  before_action :sync_accounts
  after_action :set_user_last_login
end
```

Another approach is to move everything into a cron job that runs once a day.

zachgoll replied:
What about lifting up authentication to happen at the application controller level? (and ignoring session routes)

That would probably require a separate PR, but I'm just trying to think through the best long-term option here.

I agree that we should avoid doing this in the authentication module.

zachgoll replied:
Authentication is now centralized per #598 

Let me know if you want to tackle this.  If not, happy to create a separate issue for it!

mattia-malnis replied:
I saw your PR, seems all good!
I'd like to tackle this, just a recap:

- We need to add `last_login_at` to the users table;
- At the `application_controller` level we sync user's accounts with the method `sync_accounts` above.

Let me know if I'm missing anything!

zachgoll replied:
Great, will keep this PR open—feel free to request another review when you're at a good stopping point.

And yes, that logic makes sense to me overall. The biggest focus I think is to get a "sketch" of the overall flow in place that will be easy to iterate on as we add more and more to the sync logic and process.  For example, #596 will cover some of those comments you had mentioned about global sync alerts, and we'll also be adding in error handling for the sync process in the future too.

mattia-malnis replied:
You can find the updated code, I think that's all good now!
